chrome/content/zotero/xpcom/file.js

```diff
@@ -446,6 +446,68 @@ Zotero.File = new function(){
 		});
 	};
 
+	/**
+	 * Asynchronously writes data from an nsIAsyncInputStream to a file.
+	 *
+	 * Designed to handle input streams where data may not be
+	 * immediately or fully available, such as network streams.
+	 *
+	 * @param {nsIInputStream} inputStream - The input stream to read from. This
+	 *        stream should implement nsIAsyncInputStream.
+	 * @param {string} path - The file path where the data will be written.
+	 * @param {number} byteCount - The expected number of bytes to write.
+	 *
+	 * @returns {Promise<number>} A promise that resolves with the number of bytes
+	 *          written when the operation is complete, or rejects with an error
+	 *          if any issues occur during reading or writing.
+	 */
+	this.putNetworkStream = async function (path, stream, byteCount) {
```

Comment:
Why do we need this when `putContentsAsync()` already takes an `nsIInputStream`? Can we just make any modifications to that if necessary?

Reply:
Not in my attempts. `putContentsAsync()` only works with streams that have EOF, which this does not. We need to copy a specific number of bytes and then the http server code writes the response on the same channel. And I could not find a utility function in XPCOM that copies a specific number of bytes from a stream as a single function call.